### DISCUSSION TITLE
Session Resumption Fails Due to Uninitialized Session ID Context

### DISCUSSION THREAD

**Discussion by Sudheesh-PU** [microsoft/msquic#4603] [Q&A]
Description : 
When attempting to resume a connection using the resumption ticket sent by the QUIC server after a successfully established connection, the resumption process fails. Upon checking the logs, a TLS error occurs on the server side with the message: session id context is uninitialized.

client-side reports : 
```
Indicating QUIC_CONNECTION_EVENT_SHUTDOWN_INITIATED_BY_TRANSPORT [0xbebc350] [IndicateShutdownByTransport:/home/sudheesh-19185/development/vcpkg/buildtrees/ms-quic/src/v2.3.5-e5ec31ff57.clean/src/core/connection.c:1303]
```
server-side reports:

```
TLS handshake error: error:0A000115:SSL routines::session id context uninitialized, file:7.clean/submodules/openssl3/ssl/ssl_sess.c:645 [OpenSslHandshakeErrorStr:/root/sudheesh/vcpkg/buildtrees/ms-quic/src/v2.3.5-e5ec31ff57.clean/src/platform/tls_openssl.c:2007]
```
MsQuic version
 Both client and server uses v2.3.5

Affected OS
 Linux

Additional OS information
 Debian 12

Steps to Reproduce:
1. Server start the listener with settings:
```
    // Resumption settings
    settings.ServerResumptionLevel = QUIC_SERVER_RESUME_AND_ZERORTT;
    settings.IsSet.ServerResumptionLevel = TRUE;
    // Certificate credential flags
    credentials.Flags = QUIC_CREDENTIAL_FLAG_REQUIRE_CLIENT_AUTHENTICATION |
                       QUIC_CREDENTIAL_FLAG_SET_CA_CERTIFICATE_FILE |
                       QUIC_CREDENTIAL_FLAG_USE_PORTABLE_CERTIFICATES |
                       QUIC_CREDENTIAL_FLAG_INDICATE_CERTIFICATE_RECEIVED
```
2. Client is initialised with :
```
    // Certificate credential flags
    credentials.Flags =
      QUIC_CREDENTIAL_FLAG_CLIENT |
      QUIC_CREDENTIAL_FLAG_SET_CA_CERTIFICATE_FILE |
      QUIC_CREDENTIAL_FLAG_USE_TLS_BUILTIN_CERTIFICATE_VALIDATION;
``` 
3. Client initialise the connection by TLS 1.3 handshake with server.
4. Server send a new resumption ticket.
5. Client store the resumption ticket received.
6. Shutting down the the connection from client side after successfully transferring some packets.
7. Client starts new connection with the session ticket received above.
8. Server close the connection with TLS handshake error.

Expected Behavior: 
The server should successfully resume the connection using the provided resumption ticket without errors.

Actual Behavior: 
The connection fails to resume due to TLS handshake error.

**Comment by anrossi** [comment]
For step 4, could you share the server code that sends the resumption ticket?
And for step 5, could you share the client code for receiving/storing the ticket?
For step 7, could you share the client code for setting the session ticket?

Are you able to repro this with logging enabled and share the logs here? Instructions for collecting logging is here: https://github.com/microsoft/msquic/blob/main/docs/Diagnostics.md

**Reply by Sudheesh-PU** [reply]
Hi @anrossi, Thank you for your response. 

1. On step 4, server code that used for send the resumption ticket is 
```
//  Sending the resumption ticket on connection callback handler
    case QUIC_CONNECTION_EVENT_CONNECTED:
         MsQuic->ConnectionSendResumptionTicket(connection, QUIC_SEND_RESUMPTION_FLAG_NONE, 0, NULL);
         break;
```
2. On step 5, the client receives the resumption ticket and store it by
```
//  Declaring global variable for resumption ticket and ticket size
    static uint8_t *resumptionTicket = nullptr;
    static uint32_t resumptionTicketLength = 0; 

//  Receiving the resumption ticket on connection callback handler
    case QUIC_CONNECTION_EVENT_RESUMPTION_TICKET_RECEIVED: {
          resumptionTicket = new uint8_t[Event->RESUMPTION_TICKET_RECEIVED.ResumptionTicketLength];
          resumptionTicketLength = Event->RESUMPTION_TICKET_RECEIVED.ResumptionTicketLength;
          memcpy(resumptionTicket, Event->RESUMPTION_TICKET_RECEIVED.ResumptionTicket, Event->RESUMPTION_TICKET_RECEIVED.ResumptionTicketLength);
    } break;
```
3. The next time, when the client trying connect it will set the resumption ticket.
```
//  Setting the resumption ticket before connectionStart() invoked
    if(resumptionTicket) {
          if (QUIC_FAILED(Status = MsQuic->SetParam(connection, QUIC_PARAM_CONN_RESUMPTION_TICKET, (uint16_t)resumptionTicketLength, resumptionTicket))) {
            throw std::runtime_error("failed to load resumption ticket");
          }
    }
//before connection handshake initiating, some data is buffered for send before the connection is started so that it can be part of the 0-RTT data.
```

NOTE : we can't share the whole source code.

**Comment by anrossi** [comment]
Thank you for sharing the code snippets. Those appear correct to me.
Are you able to collect diagnostic traces and share those with me?

**Reply by Sudheesh-PU** [reply]
These are collected diagnostic traces:  

[client.log](https://github.com/user-attachments/files/17488584/client.log)
[server.log](https://github.com/user-attachments/files/17488585/server.log)

NOTE :  I have edited the log files by changing the server IP to x.x.x.x.

Please let me know if you need any additional details or if there's anything else I should check.
